src/webresource-handler/handlers/S3Handler.ts

```diff
@@ -122,6 +131,83 @@ export class S3Handler implements WebResourceHandler {
 		return webResource;
 	}
 
+	public async getUploadUrl(
+		uploadParameters: WebResourceUploadParameters,
+		metadata: WebResourceUploadPayload['metadata'],
+	): Promise<WebResourceUploadResponse> {
+		const fileKey = this.getFileKey(
+			uploadParameters.fieldName,
+			metadata.filename,
+			true,
+		);
+		const [presignedUrl, token] = await Promise.all([
+			getSignedUrl(
+				this.client,
+				new PutObjectCommand({
+					Bucket: this.bucket,
+					Key: fileKey,
+					ContentLength: metadata.size,
+				}),
+				{
+					expiresIn: this.signedUrlExpireTimeSeconds,
+				},
+			),
+			this.generateMetadataJwt(fileKey, uploadParameters, metadata),
+		]);
+
+		return {
+			presignedUrl,
+			token,
+		};
+	}
+
+	public async decodeUploadToken(
+		token: string,
+	): Promise<WebResourceTokenPayload> {
+		return jsonwebtoken.verify(
+			token,
+			this.jwtSigningKey,
+		) as WebResourceTokenPayload;
+	}
+
+	public async commitUpload({
+		fileKey,
+		metadata,
+	}: WebResourceTokenPayload): Promise<WebResource> {
+		const newKey = fileKey.substring(`${this.stagingFolder}`.length);
+		const command = new CopyObjectCommand({
+			Bucket: this.bucket,
+			CopySource: `${this.bucket}/${fileKey}`,
+			Key: newKey,
+		});
+		await this.client.send(command);
+		return {
+			href: this.getS3URL(newKey),
+			filename: metadata.filename,
+			size: metadata.size,
+			content_type: metadata.content_type,
+		};
+	}
+
+	private async generateMetadataJwt(
+		fileKey: string,
+		uploadParameters: WebResourceUploadParameters,
+		metadata: WebResourceUploadPayload['metadata'],
+	): Promise<string> {
+		// Implement JWT generation here
+		return jsonwebtoken.sign(
+			{
+				fileKey,
+				uploadParameters,
+				metadata,
+			},
+			this.jwtSigningKey,
+			{
+				expiresIn: this.signedUrlExpireTimeSeconds,
```

Comment:
I am not sure if using the same expire timeout for both the token and the signed URL is a problem. Considering this would be slightly long (24h or so) timeouts I suppose it is fine